pkg/kubelet/custommetrics/impl.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package custommetrics
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+)
+
+const (
+	// Key for the annotation defining custom metrics for pod.
+	// More info: http://releases.k8s.io/HEAD/docs/proposals/custom-metrics.md#api
+	CustomMetricsAnnotationKey = "metrics.alpha.kubernetes.io/custom-endpoints"
+
+	CustomMetricsDefinitionContainerPath = "/etc/cadvisor-custom-metrics-prometheus.json"
+)
+
+// CustomMetricsEndpoint is a structure defining endpoint for pod's custom metric.
+// It is parsed from JSON from the value of the annotation with CustomMetricsAnnotationKey.
+// More info: http://releases.k8s.io/HEAD/docs/proposals/custom-metrics.md#api
+type CustomMetricsEndpoint struct {
+	API   string   `json:"api",omitempty`
+	Path  string   `json:"path",omitempty`
+	Port  int      `json:"port"`
+	Names []string `json:"names"`
+}
+
+func AddCustomMetricsLabelIfRequired(pod *api.Pod, labels map[string]string) {
+	if _, ok := pod.Annotations[CustomMetricsAnnotationKey]; ok {
+		labels["io.cadvisor.metric.prometheus"] = CustomMetricsDefinitionContainerPath
+	}
+}
+
+func MakeCustomMetricsMountIfRequired(
+	pod *api.Pod, container *api.Container, podContainerDir string, maxCustomMetricsPerPod int) (*kubecontainer.Mount, error) {
+	if definition, ok := pod.Annotations[CustomMetricsAnnotationKey]; ok {
+		customMetricsMount, err := makeCustomMetricsMount(
+			container, podContainerDir, pod.Status.PodIP, definition, maxCustomMetricsPerPod)
+		if err != nil {
+			return nil, err
+		}
+		return customMetricsMount, nil
+	}
+	return nil, nil
+}
+
+func makeCustomMetricsMount(container *api.Container, podContainerDir, podIP, definition string, maxCustomMetricsPerPod int) (*kubecontainer.Mount, error) {
+	endpoints, err := getCustomMetricsEndpointsForContainer(container, definition, maxCustomMetricsPerPod)
+	if err != nil {
+		return nil, err
+	}
+	if len(endpoints) == 0 {
+		return nil, nil
+	}
+
+	if err := os.MkdirAll(podContainerDir, 0750); err != nil {
+		return nil, fmt.Errorf("error creating pod-container directory: %v", err)
+	}
+	hostsFilePath := path.Join(podContainerDir, "cadvisor-custom-metrics-prometheus.json")
```

Comment:
`cadvisor-custom-metrics-prometheus.json` needs to be a constant too since it's part of `CustomMetricsDefinitionContainerPath`